File: DD_c_plusplus/Examples/IndependentSetInstance/IndependentSetInstance.cpp
#include "IndependentSetInstance.h"

IndependentSetInstance::IndependentSetInstance(const string& file_path) {
    read_file(file_path);
}

IndependentSetInstance::IndependentSetInstance(dynamic_bitset<>* initial_state,
                                               vector<pair<string, vector<int>>> variables,
                                               map<string, vector<int>> dict_node_neighbors)
    : initial_state(initial_state),
      variables(std::move(variables)),
      dict_node_neighbors(std::move(dict_node_neighbors))
{}

void IndependentSetInstance::read_file(const string &file_path) {
    ifstream input(file_path);
    if( !input.is_open() ) {
        cerr << "Error: could not open file " << file_path << endl;
        exit(1);
    }

    int variable_length;

    input >> variable_length;
    for (int i = 1; i <= variable_length; ++i) {
        string variable_name = "x_" + to_string(i);
        variables.emplace_back(make_pair(variable_name, vector<int>{0,1}));
    }

    initial_state = new dynamic_bitset<>(variable_length);
    for (int i = 0; i < variable_length; ++i) {
        initial_state->set(i);
    }


    for (int i = 0; i < variable_length; i++) {
        double value;
        input >> value;
        objective_weights.push_back(value);
    }

    for (int i = 0; i < variable_length; ++i) {
        dict_node_neighbors["x_" + to_string(i+1)] = vector<int>();
    }

    string line;
    getline(input, line);

    for (int i = 0; i <= variable_length; i++) {
        getline(input, line);
        stringstream ss(line);
        string key;
        char colon;
        ss >> key >> colon;

        key.pop_back();

        vector<int> values;
        int value;
        char comma;

        while (ss >> value >> comma) {
            values.push_back(value);
        }
        ss >> value;

        dict_node_neighbors[key] = values;

        int child_node = stoi(key.substr(2));

        for (int parent_node : values) {
            if (child_node < parent_node) {
                edges.emplace_back(child_node, parent_node);
            }
        }
    }

}

File: DD_python/Examples/IndependentSetInstance/IndependentSetInstance.py
from bitarray import bitarray

class IndependentSetStructure:
    def __init__(self, file_path: str) -> None:
        """
        SourceCode to store the structure of the independent set problem
        """
        self.variables: list[tuple[str, list[int]]] = None
        self.objective_weights: list[int] = None
        self.dict_node_neighbors: dict[str, list[int]] = None
        self.initial_state: 'bitarray' = None
        self.edges: list[tuple[int, int]] = None

        self.read_file(file_path)
    
    def read_file(self, file_path: str) -> None:
        try:
            with open(file_path, 'r') as input_file:
                variable_length: int = int(input_file.readline().strip())
                self.variables: list[tuple[str, list[int]]] = [("x_" + str(i + 1), [0, 1]) for i in range(variable_length)]

                self.initial_state = bitarray(variable_length)
                self.initial_state.setall(1)

                self.objective_weights: list[tuple[str, list[int]]] = list(map(int, input_file.readline().strip().split()))

                self.dict_node_neighbors: dict[str, list[int]] = {f"x_{i+1}": [] for i in range(variable_length)}
                self.edges: list[tuple[int, int]] = []

                for _ in range(variable_length):
                    line = input_file.readline().strip()
                    parts = line.split(':')
                    key = parts[0].rstrip()

                    values = list(map(int, parts[1][2:-1].split(',')))
                    self.dict_node_neighbors[key] = values

                    child_node = int(key[2:]) 
                    for parent_node in values:
                        if child_node < parent_node:
                            self.edges.append((child_node, parent_node))

        except IOError:
            print(f"Error: could not open file {file_path}")
            exit(1)
    
    def __str__(self):
        return f"IndependentSetStructure; variables:{self.variables}, objective_weights:{self.objective_weights}, dict_node_neighbors:{self.dict_node_neighbors}, initial_stet:{self.initial_state}, edges:{self.edges}"